[Doc] Add scheduling doc (#23343)
ray-project/ray: doc/source/ray-core/actors/scheduling.rst

Given the specified resource requirements, a node is available (has the available resources to run the actor now),
feasible (has the resources but they are not available now)
or infeasible (doesn't have the resources). If there are available nodes, Ray will choose one based on other factors discussed below.
If there are no available nodes but only feasible ones, Ray will wait until resources are freed up and nodes become available.
If all nodes are infeasible, the actor cannot be scheduled until feasible nodes are added to the cluster.

Placement Group
---------------
If ``placement_group`` option is set then the actor will be scheduled where the placement group is located.
See :ref:`Placement Group <ray-placement-group-doc-ref>` for more details.

Scheduling Strategy
-------------------
Actor supports ``scheduling_strategy`` option to specify the strategy used to decide the best node among available nodes.
Currently the supported strategy is "DEFAULT".
"DEFAULT" is the default strategy used by Ray. With the current implementation, Ray will try to pack actors on nodes
until the resource utilization is beyond a certain threshold and spread actors afterwards.

> Currently Ray handles actors that don't require any resources specially by randomly choosing a node in the cluster without considering resource utilization.

Reviewer: E.g., num_cpus=0 or not specifying resources

Author: Not specifying resources means 1 cpu is used for scheduling so the node is not chosen randomly.

Reviewer: I see. This is subtle, you should call out explicitly: currently, the only way to achieve spread scheduling for actors is to assign num_cpus=0.

Author: Yea, this is the sign that we should really simplify this.

Resources
---------
Each actor has the :ref:`specified resource requirements <resource-requirements>`.
By default, actors require 1 CPU for scheduling and 0 CPU for running.
This means they cannot get scheduled on a zero-cpu node, but an infinite number of them
can run on any non-zero cpu node. If resources are specified explicitly, they are required
for both scheduling and running.
Given the specified resource requirements, a node is available (has the available resources to run the actor now),
feasible (has the resources but they are not available now)
or infeasible (doesn't have the resources). If there are available nodes, Ray will choose one based on other factors discussed below.
If there are no available nodes but only feasible ones, Ray will wait until resources are freed up and nodes become available.
If all nodes are infeasible, the actor cannot be scheduled until feasible nodes are added to the cluster.

Placement Group
---------------
If ``placement_group`` option is set then the actor will be scheduled where the placement group is located.
See :ref:`Placement Group <ray-placement-group-doc-ref>` for more details.

Scheduling Strategy
-------------------

> Actor supports ``scheduling_strategy`` option to specify the strategy used to decide the best node among available nodes.

Reviewer: ```suggestion
Actors support ``scheduling_strategy`` option to specify the strategy used to decide the best node among available nodes.
```

> Currently the supported strategy is "DEFAULT".

Reviewer: ```suggestion
Currently the only supported strategy for actors is "DEFAULT".
```

"DEFAULT" is the default strategy used by Ray. With the current implementation, Ray will try to pack actors on nodes
until the resource utilization is beyond a certain threshold and spread actors afterwards.

> Currently Ray handles actors that don't require any resources specially by randomly choosing a node in the cluster without considering resource utilization.

Reviewer: ```suggestion
Currently Ray handles actors that don't require any resources (i.e., that specify ``num_cpus=0``) specially by randomly choosing a node in the cluster without considering resource utilization.
```

Author: This is a bit misleading since `num_cpus=0` itself is not sufficient, the actor also cannot require other resources in order to be random.

Reviewer: It's better than not specifying. You can also clarify by saying "with no other resources".

It was unclear to me what this sentence meant without this clarification.

> Since nodes are randomly chosen, actors that don't require any resources are spreaded across the cluster.

Reviewer: ```suggestion
Since nodes are randomly chosen, actors that don't require any resources are effectively SPREAD across the cluster.
```

.. tabbed:: Python

    .. code-block:: python

        @ray.remote
        class Actor:
            pass

>         # "DEFAULT" scheduling strategy is used.

Reviewer: ```suggestion
        # "DEFAULT" scheduling strategy is used (packed onto nodes).
```

Author: I already described the behavior of "DEFAULT" before, do we need to repeat here? Also it's not just pack, it's pack before a threshold and then spread.

Reviewer: Typically, redundancy is good for user understanding

        a1 = Actor.remote()

>         # Node is chosen randomly because this actor requires no resources.

Reviewer: ```suggestion
        # Zero-CPU actors are randomly assigned to nodes.
```

Author: I actually want to avoid zero-cpu actors: for example, zero-cpu, one-gpu actor is not random since it requires some resources. Many be we should call it `empty resource actors`?

Reviewer: How about zero (and no other resources). Empty is an ambiguous term, users might think leaving the decorator empty will work, but it doesn't!
